spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/aggregation/ScriptOperators.java

```diff
@@ -0,0 +1,412 @@
+		 * @return new instance of {@link Function}.
+		 */
+		public Function args(Object... args) {
+			return args(Arrays.asList(args));
+		}
+
+		/**
+		 * Set the arguments passed to the function body.
+		 *
+		 * @param args the arguments passed to the function body. Leave empty if the function does not take any arguments.
+		 * @return new instance of {@link Function}.
+		 */
+		public Function args(List<Object> args) {
+
+			Assert.notNull(args, "Args must not be null! Use an empty list instead.");
+			return new Function(appendAt(1, Fields.ARGS.toString(), args));
+		}
+
+		/**
+		 * The language used in the body.
+		 *
+		 * @param lang must not be {@literal null} nor empty.
+		 * @return new instance of {@link Function}.
+		 */
+		public Function lang(String lang) {
+
+			Assert.hasText(lang, "Lang must not be null nor emtpy! The default would be 'js'.");
+			return new Function(appendAt(2, Fields.LANG.toString(), lang));
+		}
+
+		@Nullable
+		List<Object> getArgs() {
+			return get(Fields.ARGS.toString());
+		}
+
+		String getBody() {
+			return get(Fields.BODY.toString());
+		}
+
+		String getLang() {
+			return get(Fields.LANG.toString());
+		}
+
+		@Override
+		protected String getMongoMethod() {
+			return "$function";
+		}
+
+		enum Fields {
+
+			BODY, ARGS, LANG;
+
+			@Override
+			public String toString() {
+				return name().toLowerCase();
+			}
+		}
+	}
+
+	/**
+	 * {@link Accumulator} defines a custom aggregation
+	 * <a href="https://docs.mongodb.com/master/reference/operator/aggregation/accumulator/">$accumulator operator</a>,
+	 * one that maintains its state (e.g. totals, maximums, minimums, and related data) as documents progress through the
+	 * pipeline, in JavaScript.
+	 * <p />
+	 * <code class="java">
+	 * {
+	 *   $accumulator: {
+	 *     init: ...,
+	 *     intArgs: ...,
+	 *     accumulate: ...,
+	 *     accumulateArgs: ...,
+	 *     merge: ...,
+	 *     finalize: ...,
+	 *     lang: "js"
+	 *   }
+	 * }
+	 * </code>
+	 * <p />
+	 * {@link Accumulator} can be used as part of {@link GroupOperation $group}, {@link BucketOperation $bucket} and
+	 * {@link BucketAutoOperation $bucketAuto} pipeline stages. <br />
+	 * <b>NOTE:</b> <a href="https://docs.mongodb.com/master/core/server-side-javascript/">Server-Side JavaScript</a>
+	 * execution must be
+	 * <a href="https://docs.mongodb.com/master/reference/configuration-options/#security.javascriptEnabled">enabled</a>
+	 *
+	 * @see <a href="https://docs.mongodb.com/master/reference/operator/aggregation/accumulator/">MongoDB Documentation:
+	 *      $accumulator</a>
+	 * @since 3.1
+	 */
+	public static class Accumulator extends AbstractAggregationExpression {
+
+		private Accumulator(Map<String, Object> value) {
+			super(value);
+		}
+
+		/**
+		 * Create a new {@link Accumulator}.
+		 *
+		 * @return new instance of {@link Accumulator}.
+		 */
+		public static Accumulator accumulator() {
```

Comment:
Would it make sense to have `Accumulator` being used in a builder style? Calling `Accumulator.accumulator()` basically creates a ready-to-use `AggregationExpression` without having values for all fields.

When used fluently, this doesn't make a difference, but having an object that pretends to be ready to use opens up for wrong usage.